File: ThirdPartyAdapters/vungle/vungle/src/main/java/com/vungle/mediation/VungleInterstitialAdapter.java
package com.vungle.mediation;

import static com.google.ads.mediation.vungle.VungleMediationAdapter.ERROR_AD_ALREADY_LOADED;
import static com.google.ads.mediation.vungle.VungleMediationAdapter.ERROR_BANNER_SIZE_MISMATCH;
import static com.google.ads.mediation.vungle.VungleMediationAdapter.ERROR_DOMAIN;
import static com.google.ads.mediation.vungle.VungleMediationAdapter.ERROR_INVALID_SERVER_PARAMETERS;
import static com.google.ads.mediation.vungle.VungleMediationAdapter.KEY_APP_ID;
import static com.google.ads.mediation.vungle.VungleMediationAdapter.TAG;

import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.ads.mediation.vungle.VungleBannerAd;
import com.google.ads.mediation.vungle.VungleInitializer;
import com.google.ads.mediation.vungle.util.ErrorUtil;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.MediationUtils;
import com.google.android.gms.ads.mediation.MediationAdRequest;
import com.google.android.gms.ads.mediation.MediationBannerAdapter;
import com.google.android.gms.ads.mediation.MediationBannerListener;
import com.google.android.gms.ads.mediation.MediationInterstitialAdapter;
import com.google.android.gms.ads.mediation.MediationInterstitialListener;
import com.vungle.ads.AdConfig;
import com.vungle.ads.BaseAd;
import com.vungle.ads.BaseAdListener;
import com.vungle.ads.InterstitialAd;
import com.vungle.ads.VungleException;

import java.util.ArrayList;


/**
 * A {@link MediationInterstitialAdapter} used to load and show Vungle interstitial ads using Google
 * Mobile Ads SDK mediation.
 */
@Keep
public class VungleInterstitialAdapter
        implements MediationInterstitialAdapter, MediationBannerAdapter, BaseAdListener {

    private MediationInterstitialListener mMediationInterstitialListener;
    private VungleManager mVungleManager;
    private AdConfig mAdConfig;
    private String mPlacement;

    // banner/MREC
    private MediationBannerListener mMediationBannerListener;
    private VungleBannerAdapter vungleBannerAdapter;
    private InterstitialAd ad;

    @Override
    public void requestInterstitialAd(@NonNull Context context,
                                      @NonNull MediationInterstitialListener mediationInterstitialListener,
                                      @NonNull Bundle serverParameters, @NonNull MediationAdRequest mediationAdRequest,
                                      @Nullable Bundle mediationExtras) {
        String appID = serverParameters.getString(KEY_APP_ID);
        if (TextUtils.isEmpty(appID)) {
            AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
                    "Missing or invalid App ID.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mediationInterstitialListener.onAdFailedToLoad(VungleInterstitialAdapter.this, error);
            return;
        }

        mMediationInterstitialListener = mediationInterstitialListener;
        mVungleManager = VungleManager.getInstance();
        mPlacement = mVungleManager.findPlacement(mediationExtras, serverParameters);
        if (TextUtils.isEmpty(mPlacement)) {
            AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
                    "Failed to load ad from Vungle. Missing or Invalid Placement ID.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mMediationInterstitialListener.onAdFailedToLoad(VungleInterstitialAdapter.this, error);
            return;
        }

        VungleInitializer.getInstance()
                .updateCoppaStatus(mediationAdRequest.taggedForChildDirectedTreatment());

        AdapterParametersParser.Config config = AdapterParametersParser.parse(appID, mediationExtras);
        // Unmute full-screen ads by default.
        mAdConfig = VungleExtrasBuilder.adConfigWithNetworkExtras(mediationExtras, false);
        VungleInitializer.getInstance()
                .initialize(
                        config.getAppId(),
                        context.getApplicationContext(),
                        new VungleInitializer.VungleInitializationListener() {
                            @Override
                            public void onInitializeSuccess() {
                                loadAd();
                            }

                            @Override
                            public void onInitializeError(AdError error) {
                                if (mMediationInterstitialListener != null) {
                                    mMediationInterstitialListener
                                            .onAdFailedToLoad(VungleInterstitialAdapter.this, error);
                                    Log.e(TAG, error.getMessage());
                                }
                            }
                        });
    }

    private void loadAd() {
        ad = new InterstitialAd(mPlacement, mAdConfig);
        ad.setAdListener(this);
        if (ad.canPlayAd()) {
            mMediationInterstitialListener.onAdLoaded(VungleInterstitialAdapter.this);
            return;
        }
        ad.load();
    }

    @Override
    public void showInterstitial() {
        ad.play();
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy: " + hashCode());
        if (vungleBannerAdapter != null) {
            vungleBannerAdapter.destroy();
            vungleBannerAdapter = null;
        }
    }

    // banner
    @Override
    public void onPause() {
        Log.d(TAG, "onPause");
        if (vungleBannerAdapter != null) {
            vungleBannerAdapter.updateVisibility(false);
        }
    }

    @Override
    public void onResume() {
        Log.d(TAG, "onResume");
        if (vungleBannerAdapter != null) {
            vungleBannerAdapter.updateVisibility(true);
        }
    }

    @Override
    public void requestBannerAd(@NonNull Context context,
                                @NonNull final MediationBannerListener mediationBannerListener,
                                @NonNull Bundle serverParameters, @NonNull AdSize adSize,
                                @NonNull MediationAdRequest mediationAdRequest, @Nullable Bundle mediationExtras) {
        mMediationBannerListener = mediationBannerListener;
        String appID = serverParameters.getString(KEY_APP_ID);
        AdapterParametersParser.Config config;
        config = AdapterParametersParser.parse(appID, mediationExtras);

        if (TextUtils.isEmpty(appID)) {
            AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
                    "Failed to load ad from Vungle. Missing or invalid app ID.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mediationBannerListener.onAdFailedToLoad(VungleInterstitialAdapter.this, error);
            return;
        }

        VungleInitializer.getInstance()
                .updateCoppaStatus(mediationAdRequest.taggedForChildDirectedTreatment());

        mVungleManager = VungleManager.getInstance();

        String placement = mVungleManager.findPlacement(mediationExtras, serverParameters);
        Log.d(TAG,
                "requestBannerAd for Placement: " + placement + " ### Adapter instance: " + this
                        .hashCode());

        if (TextUtils.isEmpty(placement)) {
            AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
                    "Failed to load ad from Vungle. Missing or Invalid placement ID.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mMediationBannerListener.onAdFailedToLoad(VungleInterstitialAdapter.this, error);
            return;
        }

        AdConfig adConfig = VungleExtrasBuilder.adConfigWithNetworkExtras(mediationExtras, true);
        if (!hasBannerSizeAd(context, adSize, adConfig)) {
            AdError error = new AdError(ERROR_BANNER_SIZE_MISMATCH,
                    "Failed to load ad from Vungle. Invalid banner size.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mMediationBannerListener.onAdFailedToLoad(VungleInterstitialAdapter.this, error);
            return;
        }

        // Adapter does not support multiple Banner instances playing for same placement except for
        // refresh.
        String uniqueRequestId = config.getRequestUniqueId();


        vungleBannerAdapter = new VungleBannerAdapter(placement, uniqueRequestId, adConfig,
                VungleInterstitialAdapter.this);
        Log.d(TAG, "New banner adapter: " + vungleBannerAdapter + "; size: " + adConfig.getAdSize());

        VungleBannerAd vungleBanner = new VungleBannerAd(placement, vungleBannerAdapter);

        Log.d(TAG, "Requesting banner with ad size: " + adConfig.getAdSize());
        vungleBannerAdapter
                .requestBannerAd(context, config.getAppId(), adSize, mMediationBannerListener);
    }

    @NonNull
    @Override
    public View getBannerView() {
        Log.d(TAG, "getBannerView # instance: " + hashCode());
        return vungleBannerAdapter.getAdLayout();
    }

    @Override
    public void adClick(@NonNull BaseAd baseAd) {
        if (mMediationInterstitialListener != null) {
            mMediationInterstitialListener.onAdClicked(VungleInterstitialAdapter.this);
        }
    }

    @Override
    public void adEnd(@NonNull BaseAd baseAd) {
        if (mMediationInterstitialListener != null) {
            mMediationInterstitialListener.onAdClosed(VungleInterstitialAdapter.this);
        }
    }

    @Override
    public void adImpression(@NonNull BaseAd baseAd) {

    }

    @Override
    public void adLoaded(@NonNull BaseAd baseAd) {
        if (mMediationInterstitialListener != null) {
            mMediationInterstitialListener.onAdLoaded(VungleInterstitialAdapter.this);
        }
    }

    @Override
    public void adStart(@NonNull BaseAd baseAd) {
        if (mMediationInterstitialListener != null) {
            mMediationInterstitialListener.onAdOpened(VungleInterstitialAdapter.this);
        }
    }

    @Override
    public void error(@Nullable BaseAd baseAd, @Nullable VungleException e) {
        AdError error = ErrorUtil.getAdError(e);
        Log.e(TAG, error.getMessage());
        if (mMediationInterstitialListener != null) {
            mMediationInterstitialListener.onAdClosed(VungleInterstitialAdapter.this);
        }
    }

    @Override
    public void onAdLeftApplication(@NonNull BaseAd baseAd) {
        if (mMediationInterstitialListener != null) {
            mMediationInterstitialListener.onAdLeftApplication(VungleInterstitialAdapter.this);
        }
    }

    public boolean hasBannerSizeAd(Context context, AdSize adSize, AdConfig adConfig) {
        ArrayList<AdSize> potentials = new ArrayList<>();
        com.vungle.ads.AdSize[] sizes = new com.vungle.ads.AdSize[] {
                com.vungle.ads.AdSize.VUNGLE_MREC,
                com.vungle.ads.AdSize.BANNER,
                com.vungle.ads.AdSize.BANNER_SHORT,
                com.vungle.ads.AdSize.BANNER_LEADERBOARD
        };

        for (com.vungle.ads.AdSize size : sizes) {
            potentials.add(new AdSize(size.getWidth(), size.getHeight()));
        }
        AdSize closestSize = MediationUtils.findClosestSize(context, adSize, potentials);
        if (closestSize == null) {
            Log.i(TAG, "Not found closest ad size: " + adSize);
            return false;
        }
        Log.i(TAG, "Found closest ad size: " + closestSize + " for requested ad size: " + adSize);

        if (closestSize.getWidth() == com.vungle.ads.AdSize.BANNER_SHORT.getWidth()
                && closestSize.getHeight() == com.vungle.ads.AdSize.BANNER_SHORT.getHeight()) {
            adConfig.setAdSize(com.vungle.ads.AdSize.BANNER_SHORT);
        } else if (closestSize.getWidth() == com.vungle.ads.AdSize.BANNER.getWidth()
                && closestSize.getHeight() == com.vungle.ads.AdSize.BANNER.getHeight()) {
            adConfig.setAdSize(com.vungle.ads.AdSize.BANNER);
        } else if (closestSize.getWidth() == com.vungle.ads.AdSize.BANNER_LEADERBOARD.getWidth()
                && closestSize.getHeight() == com.vungle.ads.AdSize.BANNER_LEADERBOARD.getHeight()) {
            adConfig.setAdSize(com.vungle.ads.AdSize.BANNER_LEADERBOARD);
        } else if (closestSize.getWidth() == com.vungle.ads.AdSize.VUNGLE_MREC.getWidth()
                && closestSize.getHeight() == com.vungle.ads.AdSize.VUNGLE_MREC.getHeight()) {
            adConfig.setAdSize(com.vungle.ads.AdSize.VUNGLE_MREC);
        }
        return true;
    }
}

File: ThirdPartyAdapters/vungle/vungle/src/main/java/com/vungle/mediation/VungleInterstitialAdapter.kt
//package com.vungle.mediation
//
//import android.content.Context
//import android.util.Log
//import android.view.View
//import androidx.annotation.Keep
//import com.google.ads.mediation.vungle.VungleInitializer
//import com.google.android.gms.ads.AdSize
//import com.vungle.warren.AdConfig
//
///**
// * A [MediationInterstitialAdapter] used to load and show Vungle interstitial ads using Google
// * Mobile Ads SDK mediation.
// */
//@Keep
//class VungleInterstitialAdapter : MediationInterstitialAdapter, MediationBannerAdapter {
//    private var mMediationInterstitialListener: MediationInterstitialListener? = null
//    private var mVungleManager: VungleManager? = null
//    private var mAdConfig: AdConfig? = null
//    private var mPlacement: String? = null
//
//    // banner/MREC
//    private var mMediationBannerListener: MediationBannerListener? = null
//    private var vungleBannerAdapter: VungleBannerAdapter? = null
//    override fun requestInterstitialAd(
//        context: Context,
//        mediationInterstitialListener: MediationInterstitialListener,
//        serverParameters: Bundle, mediationAdRequest: MediationAdRequest,
//        mediationExtras: Bundle?
//    ) {
//        val appID: String = serverParameters.getString(VungleMediationAdapter.Companion.KEY_APP_ID)
//        if (TextUtils.isEmpty(appID)) {
//            if (mediationInterstitialListener != null) {
//                val error = AdError(
//                    VungleMediationAdapter.Companion.ERROR_INVALID_SERVER_PARAMETERS,
//                    "Missing or invalid App ID.", VungleMediationAdapter.Companion.ERROR_DOMAIN
//                )
//                Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//                mediationInterstitialListener.onAdFailedToLoad(
//                    this@VungleInterstitialAdapter,
//                    error
//                )
//            }
//            return
//        }
//        mMediationInterstitialListener = mediationInterstitialListener
//        mVungleManager = VungleManager.instance
//        mPlacement = mVungleManager!!.findPlacement(mediationExtras, serverParameters)
//        if (TextUtils.isEmpty(mPlacement)) {
//            val error = AdError(
//                VungleMediationAdapter.Companion.ERROR_INVALID_SERVER_PARAMETERS,
//                "Failed to load ad from Vungle. Missing or Invalid Placement ID.",
//                VungleMediationAdapter.Companion.ERROR_DOMAIN
//            )
//            Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//            mMediationInterstitialListener.onAdFailedToLoad(this@VungleInterstitialAdapter, error)
//            return
//        }
//        VungleInitializer.getInstance()
//            .updateCoppaStatus(mediationAdRequest.taggedForChildDirectedTreatment())
//        val config = AdapterParametersParser.parse(appID, mediationExtras)
//        // Unmute full-screen ads by default.
//        mAdConfig = VungleExtrasBuilder.adConfigWithNetworkExtras(mediationExtras, false)
//        VungleInitializer.getInstance()
//            .initialize(
//                config.appId,
//                context.applicationContext,
//                object : VungleInitializationListener {
//                    override fun onInitializeSuccess() {
//                        loadAd()
//                    }
//
//                    override fun onInitializeError(error: AdError?) {
//                        if (mMediationInterstitialListener != null) {
//                            mMediationInterstitialListener
//                                .onAdFailedToLoad(this@VungleInterstitialAdapter, error)
//                            Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//                        }
//                    }
//                })
//    }
//
//    private fun loadAd() {
//        if (Vungle.canPlayAd(mPlacement)) {
//            if (mMediationInterstitialListener != null) {
//                mMediationInterstitialListener.onAdLoaded(this@VungleInterstitialAdapter)
//            }
//            return
//        }
//        Vungle.loadAd(mPlacement, object : LoadAdCallback() {
//            fun onAdLoad(placementID: String?) {
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdLoaded(this@VungleInterstitialAdapter)
//                }
//            }
//
//            fun onError(placementID: String?, exception: VungleException?) {
//                val error: AdError = VungleMediationAdapter.getAdError(exception)
//                Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdFailedToLoad(
//                        this@VungleInterstitialAdapter,
//                        error
//                    )
//                }
//            }
//        })
//    }
//
//    override fun showInterstitial() {
//        Vungle.playAd(mPlacement, mAdConfig, object : PlayAdCallback() {
//            fun creativeId(creativeId: String?) {
//                // no-op
//            }
//
//            fun onAdStart(placementID: String?) {
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdOpened(this@VungleInterstitialAdapter)
//                }
//            }
//
//            fun onAdEnd(placementID: String?, completed: Boolean, isCTAClicked: Boolean) {
//                // Deprecated, no-op.
//            }
//
//            fun onAdEnd(placementID: String?) {
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdClosed(this@VungleInterstitialAdapter)
//                }
//            }
//
//            fun onAdClick(placementID: String?) {
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdClicked(this@VungleInterstitialAdapter)
//                }
//            }
//
//            fun onAdRewarded(placementID: String?) {
//                // No-op for interstitial ads.
//            }
//
//            fun onAdLeftApplication(placementID: String?) {
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdLeftApplication(this@VungleInterstitialAdapter)
//                }
//            }
//
//            fun onError(placementID: String?, exception: VungleException?) {
//                val error: AdError = VungleMediationAdapter.getAdError(exception)
//                Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//                if (mMediationInterstitialListener != null) {
//                    mMediationInterstitialListener.onAdClosed(this@VungleInterstitialAdapter)
//                }
//            }
//
//            fun onAdViewed(id: String?) {
//                // No-op.
//            }
//        })
//    }
//
//    override fun onDestroy() {
//        Log.d(VungleMediationAdapter.Companion.TAG, "onDestroy: " + hashCode())
//        if (vungleBannerAdapter != null) {
//            vungleBannerAdapter!!.destroy()
//            vungleBannerAdapter = null
//        }
//    }
//
//    // banner
//    override fun onPause() {
//        Log.d(VungleMediationAdapter.Companion.TAG, "onPause")
//        if (vungleBannerAdapter != null) {
//            vungleBannerAdapter!!.updateVisibility(false)
//        }
//    }
//
//    override fun onResume() {
//        Log.d(VungleMediationAdapter.Companion.TAG, "onResume")
//        if (vungleBannerAdapter != null) {
//            vungleBannerAdapter!!.updateVisibility(true)
//        }
//    }
//
//    override fun requestBannerAd(
//        context: Context,
//        mediationBannerListener: MediationBannerListener,
//        serverParameters: Bundle, adSize: AdSize,
//        mediationAdRequest: MediationAdRequest, mediationExtras: Bundle?
//    ) {
//        mMediationBannerListener = mediationBannerListener
//        val appID: String = serverParameters.getString(VungleMediationAdapter.Companion.KEY_APP_ID)
//        val config: AdapterParametersParser.Config
//        config = AdapterParametersParser.parse(appID, mediationExtras)
//        if (TextUtils.isEmpty(appID)) {
//            if (mediationBannerListener != null) {
//                val error = AdError(
//                    VungleMediationAdapter.Companion.ERROR_INVALID_SERVER_PARAMETERS,
//                    "Failed to load ad from Vungle. Missing or invalid app ID.",
//                    VungleMediationAdapter.Companion.ERROR_DOMAIN
//                )
//                Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//                mediationBannerListener.onAdFailedToLoad(this@VungleInterstitialAdapter, error)
//            }
//            return
//        }
//        VungleInitializer.getInstance()
//            .updateCoppaStatus(mediationAdRequest.taggedForChildDirectedTreatment())
//        mVungleManager = VungleManager.instance
//        val placement = mVungleManager!!.findPlacement(mediationExtras, serverParameters)
//        Log.d(
//            VungleMediationAdapter.Companion.TAG,
//            "requestBannerAd for Placement: $placement ### Adapter instance: " + this
//                .hashCode()
//        )
//        if (TextUtils.isEmpty(placement)) {
//            val error = AdError(
//                VungleMediationAdapter.Companion.ERROR_INVALID_SERVER_PARAMETERS,
//                "Failed to load ad from Vungle. Missing or Invalid placement ID.",
//                VungleMediationAdapter.Companion.ERROR_DOMAIN
//            )
//            Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//            mMediationBannerListener.onAdFailedToLoad(this@VungleInterstitialAdapter, error)
//            return
//        }
//        val adConfig: AdConfig =
//            VungleExtrasBuilder.adConfigWithNetworkExtras(mediationExtras, true)
//        if (!mVungleManager!!.hasBannerSizeAd(context, adSize, adConfig)) {
//            val error = AdError(
//                VungleMediationAdapter.Companion.ERROR_BANNER_SIZE_MISMATCH,
//                "Failed to load ad from Vungle. Invalid banner size.",
//                VungleMediationAdapter.Companion.ERROR_DOMAIN
//            )
//            Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//            mMediationBannerListener.onAdFailedToLoad(this@VungleInterstitialAdapter, error)
//            return
//        }
//
//        // Adapter does not support multiple Banner instances playing for same placement except for
//        // refresh.
//        val uniqueRequestId = config.requestUniqueId
//        if (!mVungleManager!!.canRequestBannerAd(placement!!, uniqueRequestId)) {
//            val error = AdError(
//                VungleMediationAdapter.Companion.ERROR_AD_ALREADY_LOADED,
//                "Vungle adapter does not support multiple banner instances for same placement.",
//                VungleMediationAdapter.Companion.ERROR_DOMAIN
//            )
//            Log.w(VungleMediationAdapter.Companion.TAG, error.getMessage())
//            mMediationBannerListener.onAdFailedToLoad(this@VungleInterstitialAdapter, error)
//            return
//        }
//        vungleBannerAdapter = VungleBannerAdapter(
//            placement, uniqueRequestId, adConfig,
//            this@VungleInterstitialAdapter
//        )
//        Log.d(
//            VungleMediationAdapter.Companion.TAG,
//            "New banner adapter: " + vungleBannerAdapter + "; size: " + adConfig.getAdSize()
//        )
//        val vungleBanner = VungleBannerAd(placement, vungleBannerAdapter)
//        mVungleManager!!.registerBannerAd(placement, vungleBanner)
//        Log.d(
//            VungleMediationAdapter.Companion.TAG,
//            "Requesting banner with ad size: " + adConfig.getAdSize()
//        )
//        vungleBannerAdapter!!
//            .requestBannerAd(context, config.appId, adSize, mMediationBannerListener)
//    }
//
//    val bannerView: View
//        get() {
//            Log.d(VungleMediationAdapter.Companion.TAG, "getBannerView # instance: " + hashCode())
//            return vungleBannerAdapter!!.adLayout!!
//        }
//}